### DISCUSSION TITLE
Mocking PnPContext using Moq

### DISCUSSION THREAD

**Discussion by sachinsatav** [pnp/pnpcore#789] [Q&A] [answered]
Hi,
I am using PnP.Core in one of my project to create new sharepoint sites and I need to write unit tests around it.
I am able to mock context factory since there is an interface for it.
`var _mockContextFactory = new Mock<IPnPContextFactory>();`

However I am not able to mock **PnPContext** since there is no public constructor. Also there is not interface which I can use for mocking. 

I looked at SDK test projects - https://github.com/pnp/pnpcore/tree/dev/src/sdk/PnP.Core.Test.Common & https://github.com/pnp/pnpcore/tree/dev/src/sdk/PnP.Core.Test but the setup is quite complicated and most of method calls used in there are only exposed to SDK solution projects. 

Can someone please help with how I can mock PnPContext? Thank you.

**Comment by jansenbe** [comment, marked as answer]
@sachinsatav : currently this is not possible as you've noticed. The internal test projects depend on the internals of PnP Core and that approach is less suitable for your needs. We don't want to allow manual PnPContext object creation, hence there's no public constructor. Would it unblock you if I add a static method on PnPContext e.g. `var mockContext = PnPContext.NewPnPContextMock();`?

**Reply by sachinsatav** [reply]
@jansenbe Thanks for your reply. Adding a public static method to create a mock would definitely help however I am not sure how we can set read only properties of that mock. Is it possible to extract an interface out of it and then may be we can use it for mocking? We are fine with any approach as long as we are able to mock PnPContext. Thanks again.

**Reply by jansenbe** [reply]
Which properties do you need to be able to set for your mocked PnPContext?

**Reply by sachinsatav** [reply]
At the moment I am looking to set Uri, Web properties for my mock.

**Reply by wonderplayer** [reply]
@jansenbe did you manage to create this static method?

**Reply by jansenbe** [reply]
@wonderplayer : nope, this did not yet happen. One can always use reflection to do this, but makes sense to have a cleaner approach. I'll create an issue from this discussion to track this ask.

**Reply by jansenbe** [reply]
@sachinsatav , @wonderplayer : I've introduced an `IPnPContext` interface to support mocking. See also #948

**Reply by jansenbe** [reply]
The interface approach works as confirmed by @wonderplayer
